### DISCUSSION TITLE
Select data from table with multiple conditions

### DISCUSSION THREAD

**Discussion by nagarajuakula** [dflib/dflib#397] [Q&A] [answered]
Hi,

I have below table and would like to select rows which satisfies multiple conditions.
Select * from table where dept = 'DEV' and id in(1,2);

Table: Employees
id name                   salary        dept
-- ----------------- --------        -----
 1 Jerry Cosin           70000.0    DEV
 2 Juliana Walewski  85000.0    DEV
 3 Joan O'Hara         101000.0  DEV
 4 Robert                  60000.0    FIN
 
 I tried below ways
1. 
DataFrame idCondition=
        DataFrame.byColumn("id").of(Series.ofInt(1,2));
DataFrame deptCondition=
        DataFrame.byColumn("dept").of(Series.of("DEV"));
connector.tableLoader("Employees").eq(idCondition).eq(deptCondition).load();

2.
DataFrame condition =
        DataFrame.byColumn("id", "dept").of(Series.ofInt(1,2), Series.of("DEV"));
connector.tableLoader("Employees").eq(condition ).load();

Note: here I am getting exception if number of inputs vary between different columns.
If I use below condition I am getting data.
condition =
        DataFrame.byColumn("id", "dept").of(Series.ofInt(1), Series.of("DEV"));

I can apply one condition first and get result data frame and apply one more condition on top of earlier result dataframe to get actual result but we have bulk data in our tables and this process become tedious so seeking alternatives.
        
Can anyone help me in this.

Thanks in advance.

**Comment by andrus** [comment]
Hi @nagarajuakula ,  here are a couple of solutions:

* `tableLoader(..)` conditions are limited by their nature. For ultimate flexibility, you can use `sqlLoader(..)` instead of `tableLoader(..)`. There's a gap in the docs that we'll address shortly, but it is somewhat self-explanatory: 

```java
connector.sqlLoader("Select * from table where dept = 'DEV' and id in(1,2)").load();
```

* This particular query can still be done with `tableLoader(..)`. Calling TableLoader `eq(..)` replaces any previously set condition and hence can not be chained. So the first approach in your example is not going to work. The second approach is correct, but as you noticed you need to supply a condition DataFrame with the same size of both columns:

```java
DataFrame condition =
      DataFrame.byColumn("id", "dept").of(Series.ofInt(1,2), Series.of("DEV", "DEV"));
```

**Reply by nagarajuakula** [reply]
Thanks @andrus for the quick reply. 

Currently doing the second solution only. Actually, we are joining multiple tables and on this result data frame trying to join other table with multiple conditions. 

Thanks
Nagaraju

**Comment by andrus** [comment, marked as answer]
If you are joining multiple tables in the DB (instead of in-memory), then the first solution (`sqlLoader(..)`) is the only one that will be applicable, as `tableLoader(..)` only allows a select from a single table.

**Reply by nagarajuakula** [reply]
Thanks for the quick reply.

**Comment by andrus** [comment]
As promised, just published `SqlLoader` docs: https://dflib.org/dflib/docs/1.x/#_sqlloader_sqlsaver
